### DISCUSSION TITLE
Research on Code Formatters Across Programming Languages

### DISCUSSION THREAD

**Discussion by Enter-tainer** [Enter-tainer/typstyle#262] [Ideas]
Research and study code formatting tools for various mainstream programming languages, including but not limited to:

- [x] Ruff (Python)
- [x] Prettier (JavaScript, TypeScript, CSS, etc.)
- [x] rustfmt (Rust)
- [ ] gofmt (Go)
- [ ] clang-format (C/C++)
- [ ] rome-js-formatter https://github.com/biomejs/biome/tree/main/crates/biome_js_formatter. Interestingly it is the formatter that wins the prettier challenge https://biomejs.dev/blog/biome-wins-prettier-challenge/

1. Understand the working principles of each tool
2. Compare features, pros, and cons of different formatters
3. Learn from their designs and implementations
4. Examine how they handle special cases and edge conditions

After completing the research, summarize findings and consider how to apply the learned knowledge to typstyle

**Comment by Enter-tainer** [comment]
## Ruff
Ruff has a [ruff_formatter](https://github.com/astral-sh/ruff/tree/main/crates/ruff_formatter) crate which is [forked](https://github.com/astral-sh/ruff/pull/2872) from [rome_formatter](https://docs.rs/biome_formatter/latest/biome_formatter/trait.Format.html). It looks like ruff is heavily inspired by rome. It has an [IR](https://docs.rs/biome_formatter/latest/biome_formatter/format_element/enum.FormatElement.html) for formatting which is similar to pretty rs. 

The initiail looking of ruff's design is similar to typstyle. Parse source code to ast, then transform ast into formatter IR, then print IR into strings. Notablely, it has special handling for comments, like [finding the source code range of every comment element](https://github.com/astral-sh/ruff/blob/757c75752e1a728669bb97a412ddf205d165b0ea/crates/ruff_python_formatter/src/lib.rs#L129). Haven't figure out how it is used yet.

Ruff [categorizes comments](https://github.com/astral-sh/ruff/blob/main/crates/ruff_python_formatter/src/comments/mod.rs#L1) into three types, leading, dangling and trailing. Python doesn't have inline comments so ruff's life is easiler than typstyle.

It also has a contribution.md describing challenges in [formatting comments](https://github.com/astral-sh/ruff/blob/757c75752e1a728669bb97a412ddf205d165b0ea/crates/ruff_python_formatter/CONTRIBUTING.md?plain=1#L327-L329). 

It looks like for each ast node struct, ruff has a format struct for it. And `ast_node.format()` will transform a ast node into the corresponding format struct. All these format structs impls a format trait, which defines how to produce formatter IR for each struct.

I feels like ruff's formatter is better than pretty rs. Maybe we should investigate more and consider switch to it or rome formatter. But before we do that we should make sure it is more powerful than current one.

**Comment by Enter-tainer** [comment]
## Prettier

I'm more familar with prettier. From what i've seen before, prettier implements [Wadler's pretty printer](https://prettier.io/docs/en/technical-details) and support more [convenient extensions](https://github.com/prettier/prettier/blob/main/commands.md). 

~~How it handles comments remains unclear.~~ It has a massive file(~1k loc) for handling comments. https://github.com/prettier/prettier/blob/main/src/language-js/comments/handle-comments.js Looks like it manually handles all cases for each ast node type.

Interestingly it can also format markdown. I wonder if it do hard line wrapping and how it handles edge cases like https://github.com/Enter-tainer/typstyle/issues/75#issuecomment-2171275781 From code, it looks like the code is written by janpanese or chinese people. The idea is interesting: newlines and spaces are inter-changeable (conditionally). And it doesn't allow breaking for certain cases. I feels like it is possible to do hard line wrapping after reading prettier's implementation. https://github.com/prettier/prettier/blob/main/src/language-markdown/print-whitespace.js

**Comment by Enter-tainer** [comment]
## Rustfmt

Rustfmt [doesn't use a format IR](https://github.com/rust-lang/rustfmt/blob/add045979b30baa978b613ab849c2645c6a698b8/Design.md?plain=1#L92). Instead, it directly [visit each ast node and produce strings](https://github.com/rust-lang/rustfmt/blob/add045979b30baa978b613ab849c2645c6a698b8/src/expr.rs#L920-L960). This gives it more flexibility. For example, rustfmt can [set differen line width limit](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#use_small_heuristics) for different construct. This is impossible in prettier or ruff.

I think we cannot learn a lot from it. Maybe it is too hard and boring to make a rustfmt-ish formatter.

**Comment by QuadnucYard** [comment]
## Review of CLI arg designs (provided by DeepSeek)

### Examples of arg designs

| **Tool**         | **Language** | **Recursive Format**          | **Check Only**       | **In-Place/Output**       | **Max Width**        | **Key Features**                                                                 | **Notable Args**                                                                 |  
|-------------------|--------------|--------------------------------|-----------------------|---------------------------|-----------------------|---------------------------------------------------------------------------------|---------------------------------------------------------------------------------|  
| **Prettier**      | JS/TS/HTML   | `prettier "src/**/*.js"`       | `--check`             | `--write` (in-place) / default stdout | `--print-width`       | Multi-language, ignores files via `.prettierignore`                             | `--no-semi`, `--single-quote`                                                   |  
| **Black**         | Python       | `black src/` (default)         | `--check --diff`      | Default in-place          | `--line-length` (88)  | Opinionated, no configurability beyond line length                             | `--skip-string-normalization`                                                   |  
| **Ruff Format**   | Python       | `ruff format .` (default)      | `ruff format --check` | Default in-place / `-` for stdout | `--line-length`      | Blazing-fast, Black-compatible, supports `pyproject.toml`                      | `--config`, `--force-exclude`                                                   |  
| **autopep8**      | Python       | `autopep8 --in-place **/*.py` (shell glob) | `--diff`              | `--in-place`              | `--max-line-length`   | Fixes PEP8 violations, less strict than Black                                   | `--aggressive`                                                                  |  
| **yapf**          | Python       | `yapf -ir .`                   | `--diff`              | `-i` (in-place)           | `--style='{column_limit: 120}'` | Highly configurable, Google/YAPF styles                                        | `--style pep8`, `--parallel`                                                    |  
| **gofmt**         | Go           | `gofmt -w ./...` (`...` = recursive) | `-d` (diff)          | `-w` (in-place) / stdout  | N/A                   | Integrated with Go toolchain, enforces Go idioms                                | `-s` (simplify code)                                                            |  
| **rustfmt**       | Rust         | `cargo fmt --all`              | `--check`             | Default in-place          | `rustfmt --config width=100` | Supports custom rules via `rustfmt.toml`                                       | `--edition` (Rust version)                                                      |  
| **ESLint**        | JS/TS        | `eslint "**/*.js"`             | `--fix-dry-run`       | `--fix` (in-place)        | `max-len` rule        | Extensible with plugins, auto-fixable rules                                     | `--rule 'indent: ["error", 4]'`                                                 |  
| **shfmt**         | Shell        | `shfmt -w -s -i 2 .`           | `-d` (diff)          | `-w` (in-place)          | N/A                   | Enforces POSIX-compliant syntax, supports simplified mode (`-s`)               | `-i` (indent size), `-ln` (shell dialect)                                       |  
| **clang-format**  | C/C++        | `find . -name '*.cpp' \| xargs clang-format -i` | `-n` (dry run) | `-i` (in-place)           | `-style="{ColumnLimit: 100}"` | Supports LLVM/Google styles, integrates with editors                           | `-fallback-style`                                                               |  
| **terraform fmt** | HCL          | `terraform fmt -recursive`     | `-check`              | Default in-place          | N/A                   | Standardizes Terraform configs, detects syntax errors                          | `-diff` (show changes)                                                          |  
| **SQLFluff**      | SQL          | `sqlfluff fix --no-safety --recurse` | `lint` (no fix)    | `fix` (in-place)          | `--max-line-length`   | Supports dialect-specific rules, auto-format complex queries                   | `--dialect postgres`                                                            |  


### **Design Patterns Summary**  
- **Recursive Handling**:  
  - Python tools like Ruff and Black use implicit directory recursion.  
  - Shell/C/C++ tools (e.g., `clang-format`) rely on `find` or glob patterns.  
- **Check vs. Fix**:  
  - Most tools use `--check` or `-d` for dry runs.  
  - Ruff’s `--check` mirrors Black’s behavior but executes significantly faster.  
- **Output Control**:  
  - Tools like `gofmt` default to stdout, while Rust/Python formatters default to in-place edits.  

### Why some formatters require **`-r`/`--recursive`** flags instead of implicitly recursing with directory paths

#### **Why `-r` Exists**  
1. **Safety First**:  
   Prevents accidental bulk changes (e.g., formatting a 10,000-file project unintentionally).  

2. **Explicit > Implicit**:  
   Forces users to **opt-in** to recursion, clarifying intent and reducing errors.  

3. **Unix Philosophy**:  
   Delegates file discovery to the shell (e.g., `find` or `**` globs) instead of baking recursion into the tool.  

4. **Flexibility**:  
   Allows mixing files/directories (e.g., `formatter file1 dir/ -r`) without recursing all paths.  

#### **Why Others Skip `-r`**  
- **Opinionated Tools** (Black, Ruff): Assume recursion is common and automate it.  
- **Glob-Aware Tools**: Use patterns like `**/*.py` to handle recursion internally.  
- **Ecosystem Norms**: Follow language-specific conventions (e.g., Go’s `./...`).  

**Trade-off**: `-r` prioritizes safety/control, while implicit recursion favors ergonomics.  

#### **Comparison Table**
| **Tool**         | Recursion Method               | Philosophy                                  |
|-------------------|---------------------------------|---------------------------------------------|
| **Black**         | Implicit (`black src/`)        | "Batteries included" – assumes recursion is common. |
| **prettier**      | `--recursive` or globs         | Safety-first, avoids accidental traversal.  |
| **gofmt**         | `./...` syntax                 | Follows Go’s toolchain conventions.         |
| **clang-format**  | `find` + `xargs`               | Unix philosophy – delegates to shell.       |
| **ruff format**   | Implicit (`ruff format .`)     | Prioritizes speed and simplicity.           |
